src/wp-includes/media.php

```diff
@@ -1974,6 +1970,35 @@ function wp_img_tag_add_loading_optimization_attrs( $image, $context ) {
 		$context
 	);
 
+	$unfiltered_decoding_val = $decoding_val;
+	$decoding_val            = ! empty( $decoding_val ) ? $decoding_val : 'async';
+
+	/**
+	 * Filters the `decoding` attribute value to add to an image. Default `async`.
+	 *
+	 * Returning a falsey value will omit the attribute.
+	 *
+	 * @since 6.1.0
+	 *
+	 * @param string|false|null $value      The `decoding` attribute value. Returning a falsey value
+	 *                                      will result in the attribute being omitted for the image.
+	 *                                      Otherwise, it may be: 'async' (default), 'sync', or 'auto'.
+	 * @param string            $image      The HTML `img` tag to be filtered.
+	 * @param string            $context    Additional context about how the function was called
+	 *                                      or where the img tag is.
+	 */
+	$decoding_val = apply_filters( 'wp_img_tag_add_decoding_attr', $decoding_val, $image, $context );
+
+	if (
+		( empty( $fetchpriority_val ) && empty( $optimization_attrs['fetchpriority'] ) )
+		&& empty( $unfiltered_decoding_val )
+		&& str_contains( $image, ' src="' )
+		&& in_array( $decoding_val, array( 'async', 'sync', 'auto' ), true )
+	) {
+		// Validate the `decoding` attribute according to the spec.
+		$image = str_replace( '<img', '<img decoding="' . esc_attr( $decoding_val ) . '"', $image );
+	}
```

Comment:
Three points of feedback on this logic:
* The whole block needs to be moved below the following `if` clause, since the filter should only be run if a `src` attribute is present. This is also to maintain the same behavior as is found in `wp_img_tag_add_decoding_attr()` today.
* The `$decoding_val` variable should not have `async` as a default, it should be used only for a potentially already present `decoding` attribute in the given `$image` tag. In other words, line 1974 should be removed.
* Related to the previous point, the `wp_img_tag_add_decoding_attr` filter should only be executed if no value is already present on the image tag, i.e. if `$decoding_val` is not empty. This is also to maintain the original behavior, where that filter wasn't executed in case of an existing attribute.

To summarize, this should behave mostly the same way as the `loading` attribute logic below, except for the `wp_lazy_loading_enabled()` consideration, since such a function does not exist for the `decoding` attribute.